Importer/CsvImporter.php

```diff
@@ -23,88 +27,241 @@ class CsvImporter extends AbstractImporter
+    public function importFile($filename, $dryRun = false)
     {
+        if (!file_exists($filename)) {
+            $this->log->addConfigurationError('File not found');
+
+            return;
+        }
+
         $csv = file_get_contents($filename);
+
+        $this->import($csv, $dryRun);
+    }
+
+    /**
+     * @param string $csv
+     * @param bool   $dryRun
+     */
+    public function import($csv, $dryRun = false)
+    {
         try {
-            return $this->import($csv, $hasHeaders, $dryRun);
-        } catch (DatabaseException $exception) {
-            $exception->setDataFile($filename);
-            throw $exception;
+            if ($this->hasHeaders() && !$this->checkHeadersAreValid($csv)) {
+                return;
+            }
+
+            $this->log->setCsv($csv);
+
+            $parsedData = $this->parser->parse($csv, $this->hasHeaders(), $this->delimiter);
+
+            $this->importData($parsedData, $this->hasHeaders(), !$dryRun);
+        } catch (\Throwable $throwable) {
+            $message = sprintf(
+                '%s: %s (uncaught exception) at %s line %d while running import.',
+                get_class($throwable),
+                $throwable->getMessage(),
+                $throwable->getFile(),
+                $throwable->getLine()
+            );
+            $this->log->addCriticalError($message);
         }
     }
 
     /**
-     * @param array $csv
-     * @param bool  $hasHeaders
-     * @param bool  $dryRun
-     *
-     * @return array|null
-     * @throws DatabaseException
+     * @param array $parsedData
+     * @param bool  $dataIsAssociativeArray
+     * @param bool  $flush
      */
-    public function import($csv, $hasHeaders = true, $dryRun = false)
+    protected function importData(array $parsedData, $dataIsAssociativeArray, $flush = true)
     {
-        if (!$this->checkHeadersAreValid($csv)) {
-            return [];
+        $entities = $this->interpreter->interpret($parsedData, $dataIsAssociativeArray);
+
+        $this->log->setImportedEntities($entities);
+
+        if (!$flush || $this->log->containErrors()) {
+            $this->detachEntities($entities);
+
+            return;
         }
 
-        $parsedData = $this->parser->parse($csv, $hasHeaders, $this->delimiter);
+        $this->persistEntities($entities);
 
-        return $this->importData($parsedData, $hasHeaders, !$dryRun);
+        try {
+            $this->entityManager->flush();
+        } catch (ORMException $exception) {
+            $message = 'Error when flushing for entity.';
+            $this->handleImporterError(new Error($message, $exception));
+        }
     }
 
     /**
-     * @param string $delimiter
+     * @param array|null $entitiesToPersist
      */
-    public function setDelimiter($delimiter)
+    protected function persistEntities(array $entitiesToPersist = null)
     {
-        $this->delimiter = $delimiter;
+        foreach ($entitiesToPersist as $entity) {
```

Comment:
If it really can be null then we have to handle it here, inside.
For `detachEntities()` - the same.